### ISSUE TITLE
OIDC backchannel logout not advertised in discovery

### ISSUE THREAD

**Issue by jinnatar** [kanidm/kanidm#2412]
### I did this
- Set up an OIDC integration with `kanidm system oauth2 create demo Demo https://demo.example.com`
- Checked it's discovery end-point with `curl https://idm.example.com/oauth2/openid/demo/.well-known/openid-configuration | jq .`

This gives the following response:
```json
{
  "issuer": "https://idm.example.com/oauth2/openid/demo",
  "authorization_endpoint": "https://idm.example.com/ui/oauth2",
  "token_endpoint": "https://idm.example.com/oauth2/token",
  "userinfo_endpoint": "https://idm.example.com/oauth2/openid/demo/userinfo",
  "jwks_uri": "https://idm.example.com/oauth2/openid/demo/public_key.jwk",
  "scopes_supported": [],
  "response_types_supported": [
    "code"
  ],
  "response_modes_supported": [
    "query"
  ],
  "grant_types_supported": [
    "authorization_code"
  ],
  "subject_types_supported": [
    "public"
  ],
  "id_token_signing_alg_values_supported": [
    "ES256"
  ],
  "token_endpoint_auth_methods_supported": [
    "client_secret_basic",
    "client_secret_post"
  ],
  "display_values_supported": [
    "page"
  ],
  "claim_types_supported": [
    "normal"
  ],
  "service_documentation": "https://kanidm.github.io/kanidm/master/integrations/oauth2.html",
  "claims_parameter_supported": false,
  "request_parameter_supported": true,
  "request_uri_parameter_supported": false,
  "require_request_uri_registration": false
}
```

### I expected the following
- As per the spec (https://openid.net/specs/openid-connect-backchannel-1_0.html):
> If the OpenID Provider supports [OpenID Connect Discovery 1.0](https://openid.net/specs/openid-connect-backchannel-1_0.html#OpenID.Discovery) [OpenID.Discovery], it uses this metadata value to advertise its support for back-channel logout:
>
>    backchannel_logout_supported
        OPTIONAL. Boolean value specifying whether the OP supports back-channel logout, with true indicating support. If omitted, the default value is false.

>  It SHOULD also register this related metadata value:
>
>    backchannel_logout_session_supported
        OPTIONAL. Boolean value specifying whether the OP can pass a sid (session ID) Claim in the Logout Token to identify the RP session with the OP. If supported, the sid Claim is also included in ID Tokens issued by the OP. If omitted, the default value is false. 

Since that value is not present in `.well-known/openid-configuration`, systems must assume it's not supported, ergo even if it's implemented which it seems to be, it cannot be used via discovery.

### Kanidm version details

- Output of `kanidm(d) version`: `kanidmd 1.1.0-rc.15`
- Are you running it in a container? If so, which image/tag?: `kanidm/server:latest`
- If not a container, how'd you install it: N/A
- Operating System / Version (On Unix please post the output of `uname -a`): `Linux example 6.1.0-16-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.67-1 (2023-12-12) x86_64 GNU/Linux`

### Any other comments
- #1184 is the PR that implemented backchannel logout, with more details in the coupled issue #613.
- #1997 seems related in adding some missing bits of discovery, but that seems to be just one part of discovery missing which does not cover the bits I encountered here.

**Comment by Firstyear** [comment]
The reason it's not advertised is because it's not currently supported :) backchannel logout in oidc is not the same as rfc7009 logout support. 

Unless I'm missing something it appears that for https://openid.net/specs/openid-connect-backchannel-1_0.html#BCSupport with discovery we don't have a way to ... well, advertise our logout token endpoint. The uri is only for dynamic clients, not discovery. So we don't have a way to expose our logout uri to clients. 

Similar, section 2.4 says that logout_tokens are issued "similar to an id token" but also doesn't show *where* or *how* they are issued. What attribute are they provided in when the id token is issued? 

As i'm reading it, this whole backchannel logout specification is very wishy-washy and doesn't actually explain how to create a valid client or server that can implement backchannel logout.

**Comment by TheRealGramdalf** [comment]
I would agree. From my (albeit limited) knowledge in OIDC, a back channel is essentially a loopback route - a channel of communication in the backend that isn't accessible to the outside world. I believe it is typically used in the context of a k8s cluster, but could also be expanded to a smaller homelab environment - rather than the Oauth2 resource server making a request to `https://auth.example.com`, it makes a request to `https://auth.lo.example.com` - which is then routed directly to Kanidm, instead of to the internet and back (or manipulated through DNS overrides/NAT loopback, as I currently do).

The idea is primarily relevant to Oauth2, which uses backchannels as a sort of MFA - the backchannel is a trusted network, with proper policies/logging enforced, so that the keys can be transported securely. If you want a better explanation, [this video](https://www.youtube.com/watch?v=rTzlF-U9Y6Y) does a better job than I can.

**Comment by jinnatar** [comment]
As a quick note from a phone belt I have time to write more: I was mislead
by the title of the original issue into believing it was already supported.
Oops, my bad.

I believe the loopback interpretation is a bit off track. What's meant here
as a back channel is that it's a request by the OP to trigger a logout at a
given RP which is delivered directly from the OP to the RP, thus not using
the front-channel via the user but a direct request to logout the given
user.

This can be used for example to signal RPs that an account has been
disabled and it's probably wise to forcibly expire and blocklist any
long-lived RP specific tokens given out to the user.

I'll read deeper into the spec and see if I can decipher some of the other
concerns.

**Comment by Firstyear** [comment]
> As a quick note from a phone belt I have time to write more: I was mislead by the title of the original issue into believing it was already supported. Oops, my bad. 

No problem at all! Oauth2/Oidc is a monster, and it's easy to get confused. In fact when I first wrote the oauth2 implementation, the spec is worded so poorly that I mistook the "resource server" and "client" as the same thing. It's not the only mistake I've made too .... 

> I believe the loopback interpretation is a bit off track. 

Agreed.

> What's meant here as a back channel is that it's a request by the OP to trigger a logout at a given RP which is delivered directly from the OP to the RP, thus not using the front-channel via the user but a direct request to logout the given user. This can be used for example to signal RPs that an account has been disabled and it's probably wise to forcibly expire and blocklist any long-lived RP specific tokens given out to the user. I'll read deeper into the spec and see if I can decipher some of the other concerns.

Yeah. That's the problem here is I can't find anything about the uri's and such that are involved and how they are meant to be agreed upon. 

Are there any applications "in the wild" that actually implement the RP side of this spec?

**Comment by TheRealGramdalf** [comment]
> Are there any applications "in the wild" that actually implement the RP side of this spec?

I believe this might be what you're looking for:
https://github.com/nextcloud/user_oidc/issues/447
https://github.com/nextcloud/user_oidc/pull/464

**Comment by jinnatar** [comment]
Thanks @TheRealGramdalf ! I searched high and low and didn't find anything "real" remotely easy to test against. The best I found was the Quarkus Java framework, but it seemed like a royal pain to test with unless someone is actually already using it. Auth0 and friends do have simple typescript example implementations in their documentation as well, but I imagine they may be too simplistic.

**Comment by Firstyear** [comment]
Right, so it's the other way around - the specification (like everything in oidc/oauth) is so poorly written that I had it confused as being the inverse.

This is a way for us as the AS to signal to the client / RP when the RP should terminate their session. That explains why I can't find anything about the URL to poke with the backchannel logout - you need to configure it when you create the client/rs.

**Comment by strobeltobias** [comment]
Are there plans to implement this feature in kanidm at some point?

It'd be great to use in combination with e.g. the aforementioned user_oidc app in nextcloud [1] or the matrix server synapse [2]. Both applications support backchannel logouts so they can end the users session on their end upon notification by the OIDC Provider.

[1] https://github.com/nextcloud/user_oidc/blob/53474451bb16d5339482dc0d567fd66b2bfe2c3b/README.md#backchannel-logout
[2] https://element-hq.github.io/synapse/v1.104/openid.html#oidc-back-channel-logout

**Comment by yaleman** [comment]
At some point yes, but not currently in development.

**Comment by johnnypea** [comment]
What is the recommended way to logout from an app using Kanidm as OIDC provider? 

Can I use `/v1/logout` endpoint to logout me user?

**Comment by candlerb** [comment]
> What is the recommended way to logout from an app using Kanidm as OIDC provider?

There isn't a general answer to that question. Each app will have its own logout endpoint. That is: each application will set its own cookie having successfully logged in via OIDC, and will have its own endpoint for clearing that cookie.

For example: say the relying party (application you've logged into) is [OAuth2-Proxy](https://oauth2-proxy.github.io/oauth2-proxy/). It has a list of [endpoints](https://oauth2-proxy.github.io/oauth2-proxy/features/endpoints), and one of those is `/oauth2/sign_out`. If you manually navigate to `myapp.com/oauth2/sign_out?rd=%2f` then it will clear your cookie and redirect you to the top page. But to logout everywhere, you'd have to do that for *every* app you've logged into using kanidm, since they'll all have their own cookies.

That's the point of the backchannel logout spec: kanidm would need to keep track of all the apps it has logged you into (see spec para 2.3), and when you logout of kanidm, it would simultaneously call out to the logout endpoints of all the apps you've visited.

But don't feel bad. Most other OIDC providers don't have this anyway, for example [authentik doesn't](https://github.com/goauthentik/authentik/issues/12710). [Zitadel claims to](https://zitadel.com/docs/guides/integrate/back-channel-logout) but I have not tested that.

Right now it's probably best to configure each app with a cookie validity of, say, 12 hours - then by the time tomorrow comes around, the cookies should have expired naturally. It's not a good idea to set it too short, as it's annoying if you frequently need to re-login via kanidm.

(Aside: I don't think that OAuth2-Proxy actually implements the RP part of the backchannel logout spec, because it's supposed to receive and validate a logout token. It unconditionally clears the cookie. However, optionally it can be configured to call a `backendLogoutURL`: that is, when you sign out of the app, it can trigger a signal the IDP to logout. The IDP would still need to make backchannel logout requests to all the *other* apps you've visited)

**Comment by Firstyear** [comment]
> > What is the recommended way to logout from an app using Kanidm as OIDC provider?
> 
> 
> That's the point of the backchannel logout spec: kanidm would need to keep track of all the apps it has logged you into (see spec para 2.3), and when you logout of kanidm, it would simultaneously call out to the logout endpoints of all the apps you've visited.

Good think we do already keep track of each oauth2 application session individually. 

> 
> But don't feel bad. Most other OIDC providers don't have this anyway, for example [authentik doesn't](https://github.com/goauthentik/authentik/issues/12710). [Zitadel claims to](https://zitadel.com/docs/guides/integrate/back-channel-logout) but I have not tested that.
> 
> Right now it's probably best to configure each app with a cookie validity of, say, 12 hours - then by the time tomorrow comes around, the cookies should have expired naturally. It's not a good idea to set it too short, as it's annoying if you frequently need to re-login via kanidm.

Well, it depends. If the cookie timeout is low, then the application should be using the *refresh* token instead of going back to kanidm. If it's having to go back to kani each time, it's the applications fault. 

We specifically implement access tokens to have short lifetimes (15 minutes) so that the application should be refreshing 'frequently' to validate the session is still valid. That way if the session is ended, the application cookie/session should reflect that in a pretty reasonable period.

But if the application isn't refreshing frequently ... well that's on the application. It should also be bounding it's session times to the access token time too ...

**Comment by kezyr** [comment]
Any update to this issue?

I add simple examples to slightly "stress up" no support for this as it creates a security issue to the user journey at all, when you use kanidm for example with nextcloud...
 
1. nextcloud with its default login page:
- user sees nextcloud login page, hits Login with Kanidm, after he logs in is redirected directly to the nextclouds dashboard and can work
- when finished, he selects Logout from dropdown menu and the default nextcloud login page is back... User thinks he is logged out, so he leaves the browser...
Security issue: user thinks he is logged out, but he is still logged in but it is not visible. When you access the nextcloud page and just hits "Login with Kanidm", the original user is back without login needed, because he logged out from nextcloud only, not from sso.
Yes you can argue - it is a users fail because he does not understand how it works. But honestly, user should not be required to understand this. For this very simple setup, it is safer to not use kanidm but to use default nextcloud login. 
 
2. nextcloud setup with allow_multiple_user_backends = 0.
- user access nextcloud url and is redirected automatically to kanidm - because why not - this should be the default setup with oidc - to not use nextcloud login
- user logs in to kanidm, is redirected to the nextcloud dashboard...
- when user hits Logout from dropdown menu, nothing happens - because under hood is redirected to kanidm, which logs in back and redirects to the dashboard. So user can think "logout does not work" because nothing happened.
This setup is better, because user at least does not think he is logged out so it solves the security issue, he will claim it is broken setup and also would prefer a way without kanidm to make "login and logout just work".

I know missing of this feature is not a security risk of kanidm. But when you think about it as a whole setup, lack of this feature is making the security issue to the user experience (1), or makes the user confused (2).

**Comment by Firstyear** [comment]
> Any update to this issue?

Again, this is not how backchannel works.

Backchannel logout allows *kanidm* to signal to *nextcloud* when it should terminate a session.

It is NOT for *nextcloud* to signal a logout event to *kanidm*. 

The direction is the opposite.

What you want is nextcloud to indicate to kanidm to end the session. This flow is VERY problematic. See this comment for more.

https://github.com/kanidm/kanidm/issues/3752#issuecomment-3130250162

**Comment by Firstyear** [comment]
I'm going to take this chance to lock this issue, as it is confusing to people. This comments serves to explain why.

**OIDC BACKCHANNEL LOGOUT IS NOT A GLOBAL SIGN OUT FEATURE**

If you are interested in global sign out, see https://github.com/kanidm/kanidm/issues/4240

OIDC backchannel logout allows Kanidm to signal to an OAuth2 client that a related user session/token has now ended and should have the session expired in the OAuth2 client. 

It's important to note *this feature is not actually even needed*. It exists to work around limitations of incorrectly coded and developed OAuth2 clients.

A correctly implemented OAuth2 Client limits the users session life to the length of the refresh token, and periodically validates the access token / exchanges the refresh token to ensure the session is still valid within Kanidm.

The intent of this is that with a short lifetime access token, a users session will frequently be checked for validity against Kanidm. If the user signs out of Kanidm, then this will immediately cause all access tokens issued as part of that session to become invalid. This leads to all OAuth2 clients seeing the users session as ended.

Incorrectly coded clients *DO NOT* link these times together - when they receive the access and refresh tokens, they issue a session token and *DO NOT* validate those tokens again periodically in line with the intent of Kanidm. This leads to a situation where when the session in Kanidm is terminated, the OAuth2 client session continues, sometimes indefinitely. 

This is what backchannel logout is for - to work around these defective clients by giving an out of band signal to terminate their sessions. 

If a client requires backchannel logout, it is not a Kanidm bug - it is a bug in the OAuth2 client implementation. They MUST correctly implement their sessions to be bound by the token lifetimes as directed by the OAuth2 server (Kanidm).
